### PR TITLE
[BC-BREAKING] Turn off scalar_check for masked_select.

### PR DESCRIPTION
**Pull request by gchanan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29955 [WIP] Turn off scalar_check for _th_normal.
* #29954 Turn off scalar_check for _th_gather.
* #29953 Turn off scalar_checks for torch.take.
* #29952 Turn off scalar_check for Tensor.set_(Tensor)
* #29951 Stop binding std_single and var_single from TH;  they aren't used anymore.
* #29950 Turn off scalar_check for set_(Storage, ...)
* #29949 Turn off scalar_check for reductions _th_max, _th_min.
* #29948 Turn off scalar_check for mode; the underlying code is correct.
* #29947 Turn off scalar_check for ormqr.
* #29946 Turn off scalar_checks for _th_multinomial_alias_draw.
* #29945 Turn off scalar_check for _th_addmv and _th_eig as they can never pass.
* **#29923 [BC-BREAKING] Turn off scalar_check for masked_select.**
* #29880 Turn off scalar_checks for __and__ and clone.
* #29879 Turn off scalar_check for __or__
* #29878 Turn off scalar_check for lshift, rshift.

Note that this changes the behavior of masked_select when both "self" and "mask" are 0-dimensional.

In previous versions of PyTorch, this would return a 0-dimensional tensor.  But the documentation reads:
"Returns a new 1-D tensor which indexes the input tensor according to the boolean mask mask which is a BoolTensor."

Code example change:
Previously:
```
>>> torch.masked_select(torch.tensor(0), torch.tensor(True))
tensor(0)
```

Now:
```
>>> torch.masked_select(torch.tensor(0), torch.tensor(True))
tensor([0])
```

Differential Revision: [D18539560](https://our.internmc.facebook.com/intern/diff/D18539560)

### CODE REVIEW

**Review by albanD — APPROVED** [review]
What is the exact change of behavior here?

**Comment by gchanan** [comment]
@albanD it's in the commit message:

"In previous versions of PyTorch, this would return a 0-dimensional tensor. But the documentation reads:
"Returns a new 1-D tensor which indexes the input tensor according to the boolean mask mask which is a BoolTensor."

So we follow the documentation now.

**Comment by albanD** [comment]
My bad, didn't saw it as github collapses all the commit messages by default...
Sounds good.
